From a2b80e9debfae7c232f70f8a5942af747401d345 Mon Sep 17 00:00:00 2001
From: Pavan S <psreenivas@habana.ai>
Date: Tue, 23 May 2023 13:05:36 +0300
Subject: [PATCH] [SW-143472] gaudi2: erom: set recovery flag in init

Due to a bug in arcboot rom in force recovery the
recovery flag is not set. This patch adds support
to set the flag in init and resets the flag before
reset. The recovery flag is used by the firmware
loader agent to allow only the fits with load type
as flash in recovery flow i.e, only firmware update
is allowed in recovery flow. The recovery flag is
cleared post successful firmware update.

Signed-off-by: Pavan S <psreenivas@habana.ai>
Change-Id: I9e3ed1daa053cb4d8aae62ae1fe861215d3883a7
---

diff --git a/app/ext_rom/erom_recovery.c b/app/ext_rom/erom_recovery.c
index 4a58c30..7e97f95 100644
--- a/app/ext_rom/erom_recovery.c
+++ b/app/ext_rom/erom_recovery.c
@@ -88,6 +88,8 @@
 	op_status.device_status = HL_I2CI_REQ_DEV_STS_RECOVERY;
 }
 
+__weak void preboot_recovery_flag(bool en) { }
+
 void erom_main_function(void)
 {
 	device_init();
@@ -107,6 +109,8 @@
 
 	hl_dev_pll_init();
 
+	preboot_recovery_flag(true);
+
 	hl_uart_init();
 
 	preboot_boot_info_init();
diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index 6198350..42103e2 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -1303,4 +1303,13 @@
 	gaudi2_ack_gic_irq();
 }
 
+void preboot_recovery_flag(bool en)
+{
+	struct gaudi2_cold_rst_data cold_rst_data;
+
+	cold_rst_data.data = regmap_readl(mmPSOC_GLOBAL_CONF_COLD_RST_FLOPS_2);
+	cold_rst_data.recovery_flag = en;
+	regmap_writel(cold_rst_data.data, mmPSOC_GLOBAL_CONF_COLD_RST_FLOPS_2);
+}
+
 void preboot_dev_spec_legacy_vers_setup(enum preboot_vers vers_type, u32 offset) {}
diff --git a/arch/gaudi2/common/reset.c b/arch/gaudi2/common/reset.c
index 5892f7c..90df013 100644
--- a/arch/gaudi2/common/reset.c
+++ b/arch/gaudi2/common/reset.c
@@ -86,6 +86,8 @@
 
 void hl_dev_reset(void)
 {
+	preboot_recovery_flag(false);
+
 	hl_set_reset_cause(HL_SW_ALL_RST);
 
 	hl_pcie_disable_access();
diff --git a/include/preboot.h b/include/preboot.h
index ef75fcd..87f7918 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -269,6 +269,7 @@
 u32 hl_get_pll_div_fix(void);
 void preboot_boot_info_init(void);
 void ppboot_boot_info_init(void);
+void preboot_recovery_flag(bool en);
 
 #endif /* __ASSEMBLY__ */
 
